pkg/kubelet/images/pullmanager/image_pull_manager.go

```diff
@@ -0,0 +1,472 @@
+	}
+
+	if !f.requireCredentialVerification(image, len(imageRef) > 0, imagePulledByKubelet) {
+		return false
+	}
+
+	if pulledRecord == nil {
+		// we have no proper records of the image being pulled in the past, we can short-circuit here
+		return true
+	}
+
+	sanitizedImage, err := trimImageTagDigest(image)
+	if err != nil {
+		klog.ErrorS(err, "failed to parse image name, forcing image credentials reverification", "image", sanitizedImage)
+		return true
+	}
+
+	cachedCreds, ok := pulledRecord.CredentialMapping[sanitizedImage]
+	if !ok {
+		return true
+	}
+
+	if cachedCreds.NodePodsAccessible {
+		// anyone on this node can access the image
+		return false
+	}
+
+	if len(cachedCreds.KubernetesSecrets) == 0 {
+		return true
+	}
+
+	for _, podSecret := range podSecrets {
+		for _, cachedSecret := range cachedCreds.KubernetesSecrets {
+
+			// we need to check hash len in case hashing failed while storing the record in the keyring
+			hashesMatch := len(cachedSecret.CredentialHash) > 0 && podSecret.CredentialHash == cachedSecret.CredentialHash
+			secretCoordinatesMatch := podSecret.UID == cachedSecret.UID &&
+				podSecret.Namespace == cachedSecret.Namespace &&
+				podSecret.Name == cachedSecret.Name
+
+			if hashesMatch {
+				if !secretCoordinatesMatch && len(cachedCreds.KubernetesSecrets) < writeRecordWhileMatchingLimit {
+					// While we're only matching at this point, we want to ensure this secret is considered valid in the future
+					// and so we make an additional write to the cache.
+					// writePulledRecord() is a noop in case the secret with the update hash already appears in the cache.
+					if err := f.writePulledRecord(image, imageRef, &kubeletconfiginternal.ImagePullCredentials{KubernetesSecrets: []kubeletconfiginternal.ImagePullSecret{podSecret}}); err != nil {
+						klog.ErrorS(err, "failed to write an image pulled record", "image", image, "imageRef", imageRef)
+					}
+				}
+				return false
+			}
+
+			if secretCoordinatesMatch {
+				if len(cachedCreds.KubernetesSecrets) < writeRecordWhileMatchingLimit { // we already know the hashes did not match, no need to check again
+					// While we're only matching at this point, we want to ensure the updated credentials are considered valid in the future
+					// and so we make an additional write to the cache.
+					// writePulledRecord() is a noop in case the hash got updated in the meantime.
+					if err := f.writePulledRecord(image, imageRef, &kubeletconfiginternal.ImagePullCredentials{KubernetesSecrets: []kubeletconfiginternal.ImagePullSecret{podSecret}}); err != nil {
+						klog.ErrorS(err, "failed to write an image pulled record", "image", image, "imageRef", imageRef)
+					}
+					return false
+				}
+			}
+		}
+	}
+
+	return true
+}
+
+func (f *PullManager) PruneUnknownRecords(imageList []string, until time.Time) {
+	f.pulledAccessors.GlobalLock()
+	defer f.pulledAccessors.GlobalUnlock()
+
+	pulledRecords, err := f.recordsAccessor.ListImagePulledRecords()
```

Comment:
if we get partial success here (pulledRecords have some entries, *and* some errors were encountered decoding), I think we should log AND continue to clean up what we can